docs/retrieval-augmented-generation/rag-framework.md

```diff
@@ -0,0 +1,72 @@
+1. Ease of use and documentation. Being strictly focused on RAG as opposed to taking a generalist approach, the learning curve of Haystack is less steep than that of Langchain. At the same time, Haystack offers extensive documentation and tutorials which are more well-organized and easy to use than those of LlamaIndex.
+1. Extending the previous point, Haystack can be seen as a middle ground between Langchain and LlamaIndex, sharing their benefits while only partially inheriting their drawbacks. Like the former, Haystack enables building custom flows and pipelines. Unlike Langchain though, Haystack does not try to be too abstract and general, strictly focusing on RAG and document search instead. As a result, Haystack is more straightforward to use, especially for users looking to implement custom and highly non-standard scenarios. On the other hand, like LlamaIndex, Haystack's performance is optimized towards data retrieval and indexing, but it offers a higher degree of flexibility and better interfaces for custom use cases.
+
+## Goals
+
+- We need to identify a framework that will be used to help support the RAG workstream. This is one of many pieces involved in the RAG workstream but identifying the framework used to index and retrieve data is step 1 of the larger picture.
+
+## Non-goals
+
+- What can we ignore when making this decision?
+
+## Decision
+
+Upon acceptance of this integration our next step is to include additional ADRs for the subsequent components required for the RAG pipeline. Next up will be decisions on Milvus and containerized/non-containerized offerings of that solution.
+
+Upon denial of this integration the team will need to go back and evaluate alternative technologies and ensure they meet the needs of the project goal and ensure they meet the larger project plan goals of a configurable RAG pipeline.
+
+## How
+
+A downstream should be generated of the [https://github.com/deepset-ai/haystack](https://github.com/deepset-ai/haystack) project.
+
+## Alternatives
+
+- Langchain/Langgraph. Drawbacks include:
+  - A generalist framework (as opposed to a RAG-focused solution)
+  - Complicated as compared to the alternatives, steep learning curve
+  - Lower performance in large-scale production environments than LlamaIndex and Haystack
+- LlamaIndex
+  - Limited flexibility and customization options as compared to the alternatives
+  - Limited out-of-the-box support for building complex, multi-component pipelines as compared to Langchain and Haystack
+  - Documentation is less well-maintained and more difficult to use as compared to the alternatives
+- RAGFlow
+  - Limited support for many of the mainstream vector DB providers
+  - Limited scalability as compared to the alternatives
+
+## Risks
+
+N/A
```

Comment:
Bringing in a new dependency always introduces risks related to building that package (and its dependencies) downstream, as well as license compliance, long term supportability, etc. Those aren't unique to this package, but it would be good to see them addressed here anyway to show that a thorough vetting was done of the package, and its dependencies.

For example, is any part of the haystack-ai dependency list GPL or AGPL? Is its license compatible with the Instructlab license?

Does it, and all of its dependencies, appear to be actively maintained by more than one maintainer?

Are those maintainers responsive to requests to review PRs, especially for CVEs?

Is the project following good practices for releasing, including tagging releases, publishing source packages, and using trusted publishing automation?

Reply:
@cooktheryan this type of analysis is fairly routine for OSPO. I'd recommend reaching out to @lhawthorn

Reply:
Thanks for your comments.

None of Haystack's dependencies (jinja2, lazy-imports, more-itertools, networkx, numpy, openai, pandas, posthog, python-dateutil, pyyaml, requests, tenacity, tqdm, typing-extensions) are GPL/AGPL. All of them are Apache, MIT, BSD or PSFL.

Haystack is licensed under Apache 2.0. What else makes it compatible or non-compatible with InstructLab?

Haystack is very [actively](https://github.com/deepset-ai/haystack/pulse/monthly) [maintained](https://github.com/deepset-ai/haystack/issues?q=is%3Aissue+is%3Aclosed) and [supported](https://github.com/deepset-ai/haystack/discussions). As for active maintenance of the dependencies I didn't check them all but most are very well-known and widely used libraries. Moreover, many of them are also on the dependency list of the alternatives to Haystack.

Haystack [releases tagged versions](https://github.com/deepset-ai/haystack/releases). They do not publish to Github Packages.
Could you please elaborate what is considered "trusted publishing automation" and whether there are any additional releasing practices we need to verify?

Reply:
That would be good information to include in the body of the document.

For trusted publisher, I mean https://docs.pypi.org/trusted-publishers/. We've recently seen an instance of a package in the AI ecosystem where that trusted publishing was set up incorrectly, which led to a third-party having the ability to not only trigger releases but to inject malicious code into _multiple versions of source packages_.  https://github.com/ultralytics/ultralytics/issues/18027 for reference.

Downstream, we're building from source packages, but if we can't trust those source packages we would not want to consume the dependency.